datalad/interface/utils.py

```diff
@@ -761,3 +765,105 @@ def filter_unmodified(content_by_ds, refds, since):
                         # or a modified path under a given directory
                         or any(m.startswith(_with_sep(p)) for p in paths_refds))]
     return keep
+
+
+def eval_results(func):
+    """Decorator for return value evaluation of datalad commands.
+
+    Note, this decorator is only compatible with commands that return
+    status dict sequences!
+
+    Two basic modes of operation are supported: 1) "generator mode" that
+    `yields` individual results, and 2) "list mode" that returns a sequence of
+    results. The behavior can be selected via the
+    `datalad.api.return-generator` configuration variable. Default is "list
+    mode".
+
+    This decorator implements common functionality for result rendering/output,
+    error detection/handling, and logging (TODO).
+
+    Result rendering/output can be triggered via the
+    `datalad.api.result-render-mode` configuration variable. Supported modes
+    are: 'json' (one object/dict per line, like git-annex), 'simple' (TODO,
+    tailored output formating provided by each command class, if any).
+
+    Error detection works by inspecting the `status` item of all result
+    dictionaries. Any occurrence of a status other than 'ok' or 'notneeded' will
+    cause an exception to be raised. TODO: The type of exception depends on the
+    nature of complete set of errors found. If all errors have a single unique
+    type, the final exception will be of this type too. In case of a heterogeneous
+    set of error a compound exception will be raised that includes information
+    on all individual errors that occurred.
+
+    Status messages will be logged automatically, by default the following
+    association of result status and log channel will be used: 'ok' (debug),
+    'notneeded' (debug), 'impossible' (warning), 'error' (error).  Logger
+    instances included in the results are used to capture the origin of a
+    status report.
+    """
+    want_generator = dlcfg.getbool('datalad.api', 'return-generator', False)
```

Comment:
I introduced dedicated keyword arguments for eval_results in #1348 and am about to integrate it in this one. So, yes, the mechanic for such dynamic tuning of behavior is already built.